### ISSUE TITLE
HyperVolume refactoring / rewriting

### ISSUE THREAD

**Issue by Corwinpro** [facebookresearch/nevergrad#366]
Currently, the existing implementation of `Multiobjective` optimization is built on top of the hypervolume algorithm, implemented in [`pyhv.py`](https://github.com/facebookresearch/nevergrad/blob/master/nevergrad/functions/multiobjective/pyhv.py). I am slightly concerned about some aspects of this code, including readability, testing, and general bugs (including fixed #346 ). In addition to that, the code is licensed under LGPL while the rest of the project is MIT.

Would it make sense to contribute to `nevergrad` and implement the same functionality (not based on the existing code, but retaining the algorithms and API) and make it MIT? I could look into that, if @teytaud and @jrapin approve this idea.

**Comment by jrapin** [comment]
I'm totally in favor of it if you're up to it! @teytaud is more knowledgeable on the this part than me though, so let's check with him also but I expect he'll be happy too ;)

**Comment by teytaud** [comment]
Also strongly in favor of it.
I admit my code is a mess.

**Comment by Corwinpro** [comment]
Thank you @teytaud , I apologise if my message made you think so. I very much appreciate your contribution to the open source community, and reckon nevergrad is the best tool for global optimization I have seen and used.

We have signed the CLA, and I will proceed with this PR asap.

Thank you!

**Comment by jrapin** [comment]
@Corwinpro thanks for the compliment ;)
If you haven't done it yet, don't hesitate to join the facebook user group if you want to: https://www.facebook.com/groups/nevergradusers/

Concerning the PR, don't hesitate to ask questions if you have any, or discuss anything that comes to mind, or ask for help.

On my side I see two parts on the multiobjective functionality:
- the core tools which are in multiobjective/pyhv.py
- the way we interface it in nevergrad which is in multiobjective/core.py

If I'm correct, you plan on only touching the first part right? Removing the external code would be great indeed!
Just so you know and in case you have any idea, I'm not very happy with the second part either for now and I think we can do much better. In particular, storing args and kwargs of the function there:
https://github.com/facebookresearch/nevergrad/blob/b1a8bff6e257b911af6cb8d611e5372087e92b24/nevergrad/functions/multiobjective/core.py#L41
is really not convenient to work with (to access first argument of first point in the front you need to write `pareto_front[0][0][0]`...), so I'm considering storing directly the `Candidate` (or in the near future the `Parameter` once our [big refactoring](https://www.facebook.com/notes/nevergrad-users/moving-to-new-parametrization-upcoming-unstability-and-breaking-changes/639090766861215/) is done) provided by nevergrad, but that would mean a tight coupling between the MultiobjectiveFunction and the rest of nevergrad, which I was trying to avoid (maybe unnecessarily).

To summarize, if you have any idea on making the whole interface cleaner, don't hesitate to let us know ;)

**Comment by jrapin** [comment]
Closing this since it is now implemented, thanks @Corwinpro
